### PR TITLE
feat(ModalWrapper): prop -> state sync for open state

### LINKED ISSUE #343: Handle to close a modal programmatically
## Detailed description

Describe in detail the issue you're having. Is this a feature request (new component, new icon), a bug, or a general issue?

feature request or could be a bug

> Is this issue related to a specific component?

yes, modals

> What did you expect to happen? What happened instead? What would you like to see changed?

A way to programmatically close the modal so that I could close it using some other control or based on some other condition.

I am not sure if handleClose prop is supposed to help in this. If it is indeed supposed to do it, probably an example of how to do it could help.

> What browser are you working in?

chrome

> What version of the Carbon Design System are you using?

carbon-components: 8.1.8
carbon-components-react: 4.1.2

> What offering/product do you work on? Any pressing ship or release dates we should be aware of?

Working on IBM App Launch service on bluemix

### PR DESCRIPTION
**Pull request by asudoh**
Fixes #343.

#### Changelog

**New**

* prop -> state sync for `open` state, for `<ModalWrapper>`
* Similar sync mechanism to:
  * `<ContentSwitcher>`
  * `<InteriorLeftNavList>`
  * `<TileGroup>`

### CODE REVIEW

File: src/components/ContentSwitcher/ContentSwitcher-test.js [hunk 1]
```diff
@@ -49,6 +54,31 @@ describe('ContentSwitcher', () => {
       expect(children.first().props().selected).toEqual(false);
       expect(children.last().props().selected).toEqual(true);
     });
+
+    it('should avoid change the selected index upon setting props, unless there the value actually changes', () => {
```
Comment (joshblack) [inline src/components/ContentSwitcher/ContentSwitcher-test.js]: Minor nit:

```
it('should avoid changing the selected index when receiving new props unless the value has changed', () => {
```

File: src/components/ContentSwitcher/ContentSwitcher.js [hunk 2]
```diff
@@ -39,6 +39,12 @@ export default class ContentSwitcher extends React.Component {
     }
   };
 
+  componentWillReceiveProps({ selectedIndex }) {
```
[the remark this answers is not in the document]
Comment (asudoh) [reply src/components/ContentSwitcher/ContentSwitcher.js]: Did you mean `gDSFP`? If so, we can definitely do for `v9`. Given we want to apply the new props for rendering, using `cDU` here will cause duplicate render logic to run.

**Comment by marijohannessen** [comment]
@asudoh can you fix the conflicts here? also did your change requests get addressed @joshblack ?

**Comment by asudoh** [comment]
Superceded by #992 - My apologies for the confusion @marijohannessen!

**Pull request closed without merging**